### ISSUE TITLE
Problems with 1.18.0 Starting and stopping

### ISSUE THREAD

**Issue by andrew-edgar** [cloudfoundry/garden-runc-release#123]
Having lots of issues with rolling an update from 1.17.1 in CFD 6.8.0 to 1.18.0.

Monit keeps flapping the garden job with the following errors ...

```
[UTC Jan 31 00:02:28] info     : 'garden' trying to restart
[UTC Jan 31 00:02:28] info     : 'garden' stop: /var/vcap/jobs/garden/bin/garden_ctl
[UTC Jan 31 00:02:39] info     : 'garden' start: /var/vcap/jobs/garden/bin/garden_ctl
[UTC Jan 31 00:02:50] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
[UTC Jan 31 00:02:50] info     : 'garden' trying to restart
[UTC Jan 31 00:02:50] info     : 'garden' stop: /var/vcap/jobs/garden/bin/garden_ctl
[UTC Jan 31 00:03:00] info     : 'garden' start: /var/vcap/jobs/garden/bin/garden_ctl
[UTC Jan 31 00:03:11] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
[UTC Jan 31 00:03:11] info     : 'garden' trying to restart
[UTC Jan 31 00:03:11] info     : 'garden' stop: /var/vcap/jobs/garden/bin/garden_ctl
[UTC Jan 31 00:03:22] info     : 'garden' start: /var/vcap/jobs/garden/bin/garden_ctl
[UTC Jan 31 00:03:33] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
[UTC Jan 31 00:03:33] info     : 'garden' trying to restart
[UTC Jan 31 00:03:33] info     : 'garden' stop: /var/vcap/jobs/garden/bin/garden_ctl
[UTC Jan 31 00:03:43] info     : 'garden' start: /var/vcap/jobs/garden/bin/garden_ctl
[UTC Jan 31 00:03:54] info     : 'garden' connection succeeded to UNIX[/var/vcap/data/garden/garden.sock]
```

hmmm very interesting.  I changed the timeout to `with timeout 10 seconds for 12 cycles` and then did a monit reload.  it now actually looks like it’s trying for the 12 times.  (it’s still failing) but I want to see what happens when it gives up and tries again.

```[UTC Jan 31 14:58:32] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
[UTC Jan 31 14:58:42] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
[UTC Jan 31 14:58:52] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
[UTC Jan 31 14:59:02] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
[UTC Jan 31 14:59:12] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
[UTC Jan 31 14:59:22] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
[UTC Jan 31 14:59:32] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
[UTC Jan 31 14:59:42] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
[UTC Jan 31 14:59:52] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
[UTC Jan 31 15:00:02] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
[UTC Jan 31 15:00:12] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
[UTC Jan 31 15:00:22] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
[UTC Jan 31 15:00:22] info     : 'garden' trying to restart
[UTC Jan 31 15:00:22] info     : 'garden' stop: /var/vcap/jobs/garden/bin/garden_ctl```
```[UTC Jan 31 15:00:22] info     : 'garden' trying to restart
[UTC Jan 31 15:00:22] info     : 'garden' stop: /var/vcap/jobs/garden/bin/garden_ctl
[UTC Jan 31 15:00:32] info     : 'garden' start: /var/vcap/jobs/garden/bin/garden_ctl
[UTC Jan 31 15:00:44] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
[UTC Jan 31 15:00:44] info     : 'garden' trying to restart
[UTC Jan 31 15:00:44] info     : 'garden' stop: /var/vcap/jobs/garden/bin/garden_ctl
[UTC Jan 31 15:01:05] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
[UTC Jan 31 15:01:05] info     : 'garden' trying to restart
[UTC Jan 31 15:01:05] info     : 'garden' stop: /var/vcap/jobs/garden/bin/garden_ctl```
so now once again it’s back into only giving it 1 try to connect.  definitely looks like a bug in monit
but also why is it taking garden so long to start up?
when I do monit stop garden.  there are a TON of “garden” processes and containerd processes still running.
it’s like the stop is not really stopping things!
## Environment

* garden-runc-release version: 1.18.0
* Enabled configuration: containerd
* IaaS: Softlayer
* Stemcell version: 170.19.0
* Kernel version: 4.15.0-43-generic

**Comment by andrew-edgar** [comment]
so we had 1.17.1 running and this was upgrading to 1.18.0.  I wonder if the issue is the "STOP" of the 1.17.1 as the monit stop does NOT seem to stop any of the processes.  Then on the start things seem to take a long time trying to cleanup and then monit gets unhappy it's taking too long.

Not sure how on one cell things all of a sudden started working 10hours later.

To get things going I had to manually kill all the garden/containerd processes and then monit start garden but this is not a sustainable method.

**Comment by andrew-edgar** [comment]
Just an update here.  I turned off containerd_mode and was able to fully run the update on the diego cells.

So this is a good thing.  It should help you look at the root cause being the slow startup or failed teardown when we have containerd in place.

**Comment by yulianedyalkova** [comment]
@andrew-edgar I just deployed garden 1.17.1 with containerd enabled and managed to succesfully update to 1.18.0. 

It is normal that you are seeing all of the garden-init, containerd-shim and dadoo processes, they are related to the running containers and survive garden/containerd restarts. 

It is weird that monit is retrying every 10 seconds and not 5 as described in the monit file, but that shouldn't create any problems. 

Do you have access to the garden/containerd logs from when the issue was still happening?

**Comment by yulianedyalkova** [comment]
I changed garden (clean deployment of 1.18.0, also happens on develop) so it sleeps for an hour before starting to listen on the socket and I noticed the following behaviour:

```
UTC Feb  1 14:22:30] info     : 'garden' start action done
UTC Feb  1 14:22:30] info     : 'garden' process is running with pid 2695
UTC Feb  1 14:22:30] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
UTC Feb  1 14:22:40] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
UTC Feb  1 14:22:50] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
UTC Feb  1 14:23:00] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
UTC Feb  1 14:23:10] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
UTC Feb  1 14:23:20] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
UTC Feb  1 14:23:30] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
UTC Feb  1 14:23:40] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
UTC Feb  1 14:23:50] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
UTC Feb  1 14:24:00] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
UTC Feb  1 14:24:10] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
UTC Feb  1 14:24:21] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
UTC Feb  1 14:24:21] info     : 'garden' trying to restart
UTC Feb  1 14:24:21] info     : 'garden' stop: /var/vcap/jobs/garden/bin/garden_ctl
UTC Feb  1 14:24:32] info     : 'garden' start: /var/vcap/jobs/garden/bin/garden_ctl
UTC Feb  1 14:24:43] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
UTC Feb  1 14:24:43] info     : 'garden' trying to restart
UTC Feb  1 14:24:43] info     : 'garden' stop: /var/vcap/jobs/garden/bin/garden_ctl
UTC Feb  1 14:24:54] info     : 'garden' start: /var/vcap/jobs/garden/bin/garden_ctl
UTC Feb  1 14:25:05] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
UTC Feb  1 14:25:05] info     : 'garden' trying to restart
UTC Feb  1 14:25:05] info     : 'garden' stop: /var/vcap/jobs/garden/bin/garden_ctl
UTC Feb  1 14:25:16] info     : 'garden' start: /var/vcap/jobs/garden/bin/garden_ctl
UTC Feb  1 14:25:27] error    : 'garden' failed, cannot open a connection to UNIX[/var/vcap/data/garden/garden.sock]
UTC Feb  1 14:25:27] info     : 'garden' trying to restart
```

It looks like the first time monit correctly tries to connect to the socket 12 times every 10 seconds, BUT after that it checks on the socket only once. So there are a few open questions:
* Why does monit ignore our 5 second interval and makes the check every 10 seconds?
* Why does it check 12 times only the first time?
* Why does it take garden so long to start (more than 2 minutes)? You said that it is the cleanup that takes long and that killing all of the garden/containerd processes unblocked everything, so it does sound a lot like https://github.com/cloudfoundry/garden-runc-release/issues/108, but it would still be nice to get more information/logs. Do you still have a cell that has the issue on it?

**Comment by andrew-edgar** [comment]
nope sorry we turned off containerd to make progress so we don't have any failing cells

**Comment by BooleanCat** [comment]
Julia and I did some experimenting with monit this morning and have been able to answer the first 2 question posed by her last week:

> Why does monit ignore our 5 second interval and makes the check every 10 seconds?

This monit configuration:

if failed unixsocket <%= p("garden.listen_address") %>
  with timeout 5 seconds for 20 cycles
  then restart

Translated to english means: Attempt to connect to the garden socket, if the connection itself takes longer than 5 seconds to establish, consider it failed. If you fail, for the next 20 cycles (a global “tick” interval defined by monit’s set daemon x, which seems to be 10 seconds on our stemcells), attempt to connect again. If you fail for 20 cycles, restart.

What’s weird is the “for 20 cycles” bit. If we allow it to attempt a connection for say 8 cycles, then monit stop garden, then monit start garden, it will make 12 more attempts. Monit seems to remember how many consecutive connection attempts made to a socket even between job restarts.

> Why does it check 12 times only the first time?

If we allow the connection check to fail for 20 times, then allow garden to start normally, we observe that the monit logs show connection successful. If we then cause garden to start failing again and restart the garden job, it never again regains that grace of 20 attempts. The 20 attempts is a global lifetime resource that can be consumed...

**Comment by yulianedyalkova** [comment]
And regarding the third question - we are getting more and more convinced that it is the issue that I linked earlier (https://github.com/cloudfoundry/garden-runc-release/issues/108) because:
* it failed on startup while trying to perform cleanup 
* we saw the exact same behaviour where hours later garden seems to have recovered (not actually, but monit process is running)
* the underlying issue happens with and without containerd enabled. The only difference is that containerd blocks on delete when that happens causing garden to fail to restart. 

Luckily, we have a fix for it in the version you are using. So you can enable containerd and you shouldn't see the problem anymore.

**Comment by andrew-edgar** [comment]
I don't understand.  we have 1.17.1 installed and are upgrading to 1.18.0 .  how can we enable containerd? because this is what caused the problem.  

We can't upgrade is the problem.   are you saying you don't need to fix anything because you have a fix already but then why didn't it work?

**Comment by andrew-edgar** [comment]
or I have to upgrade with containerd off and then turn it back on?

**Comment by BooleanCat** [comment]
We believe the issue is one that we fixed in 1.18.0. The issue will manifest itself as container processes being "forgotten" by garden and so when garden starts up again it's unable to stop that processes (garden is configured in CF to kill containers on startup).

Those container processes could have been forgotten by garden in 1.17.1 - so even when you upgrade to 1.18.0 those problem processes still exist.

There are two ways to get around this issue:

1. deploy with garden 1.18.0 with `--recreate`; or
2. delete the particular diego-cell which is having problems and deploy with garden 1.18.0

We're fairly confident this will resolve your issue (and in the unlikely event that it doesn't that's useful data).

We've taken two actions away from this issue:

1. we'll document the risk of this occuring in the release notes for garden 1.18.0
2. we're going to have a conversation with our PM about the existing behaviour of how we use monit and see what the desired behaviour should be

As for where containerd fits into this - we're not sure it's relevant. It's hard to know without the ability to poke around on the problem machine - it's feasible that the lost container processes are causing garden to lock up only in containerd mode.

**Comment by andrew-edgar** [comment]
deploy with "recreate" and delete the cell seem VERY drastic.  this is a simple upgrade scenario I need to have supported.

I need to know what the best upgrade scenario is from cfd1.19.0 with garden 1.17.2 updated to CFD 6.8.0 with garden 1.18.0.  

Right now I think I will need to make sure I am NOT using containerd (which luckily I am not in the main upgrade scenario) and so I will have more information later this week.   I can try to upgrade to 1.18.0 with containerd on to see if things are ok.  but I will not be able to get any new information until later this week

**Comment by BooleanCat** [comment]
We think there are two orthognal concerns here so let's talk about each seperately.

There was a change introduced in garden 1.18.0 that introduced an issue in how garden is deployed within cf-deployment (essentially we broke an ops file). We've merged a fix for this in cf-deployment that's not yet shipped, but to bypass the issue you can deploy with this ops file:

```
- type: remove
  path: /instance_groups/name=diego-cell/jobs/name=containerd?
```

The second issue is related to #108. This is a bug that occured as a result of systemd moving container pids out of device cgroups owned by the container (this means that runc, and therefore garden, "loses" track of which pids are inside a given container). This issue should not occur on versions of garden 1.18.0+ - however the debris left behind by deployments with earlier versions of garden may prevent the upgrade.

In this circumstance we recommend that operators recreate diego cells that have this issue (it will present itself as a deployment failure with the rep being unable to communicate with garden).

For deployments where recreates are not feasible, you can try to the following to recover a diego-cell in this state: `pkill -9 garden-init` as root on the cell after performing `monit stop garden`. If you need any help with all of this we'd be happy to work with operators to get those deployments in a good place.

**Comment by andrew-edgar** [comment]
First issue.  

fixing CFD is NOT a useful fix.  because yes you have fixed CFD but I am at CFD 6.8.0 and will not go past 6.8.0 so I need new garden versions that will work on the cfd6.8.0 level.   which means I need a new version of cfd6.8.0 that will put back a "dummy" job for containerd.  or I have to have a patchy ops file included in my code / manifest generation to remove that job. 

What is the plan for future versions of garden to support back level versions of CFD?  as garden "used" to be one of the good guys where we could upgrade garden versions without upgrading a complete CFD version.

The second one I think I can be ok with because I am upgrading from a version without containerd so I should NOT get stuck going forward.  I got stuck because I hit this in an env where I had garden 1.17.1 with containerd on.

**Comment by BooleanCat** [comment]
Hey Andrew - so I totally forgot we were fixing this is the next release (1.18.1). We just shipped this release that should allow you to perform the upgrade without maintaining any ops files to deploy it.

https://github.com/cloudfoundry/garden-runc-release/releases

**Comment by Callisto13** [comment]
Closing since this is fixed in 1.18.1. Please re-open if problem persists or if further discussion is needed.
